### DISCUSSION TITLE
sh: run-p: command not found

### DISCUSSION THREAD

**Discussion by abdipramana** [gregsullivan/_tw#67] [Q&A]
Hey guys, I following the [docs](https://underscoretw.com/docs/) and on step 4 run command `npm install && npm run dev` got error like this:
```
npm install && npm run dev
up to date, audited 1 package in 567ms
found 0 vulnerabilities
> dev
> npm run development
> development
> run-p development:**
sh: run-p: command not found
```

I use node v18.14.0 and npm 9.3.1

Is there any config I'm missing? Thanks!

**Comment by gregsullivan** [comment]
Hi Abdi—Thanks for posting your question!

Could you please confirm a few things for me:

* Which operating system are you using?
* Which folder are you located in when running these commands?
* Is there a `node_modules` folder in your repository root? If yes, does it have a large number of folder inside it?

My first impression is that something is going wrong with the install process, as it says `npm install` didn't actually install anything. (Or you've run `npm install` already, prior to this.)

If the install process has run successfully, then something about your setup is preventing your shell from running the appropriate installed applications. (`run-p` is part of `npm-run-all`, which should have been installed as part of the install process.)

**Comment by abdipramana** [comment]
Hi Greg, thanks for the response.

I'm currently use macOS Monterey, [Local](https://localwp.com/) for dev tools environment. My project structure runs commands in the `wp-content/themes/wjnt` directory, and I also have a `node_modules folder at the root of my project on `wp-content/themes/wjnt`.

After running `npm install`, I can confirm that the packages are present in the `node_modules` folder (I'll attach a screenshot for your reference). However, when I execute `npm list`, it shows an empty result and tries to clear the npm cache also.

Any suggestions or feedback on how to resolve this issue? Thank you
![Screen Shot 2023-10-19 at 10 38 43 PM](https://github.com/gregsullivan/_tw/assets/17507931/bd683181-0a84-496f-a8a3-3a536a2be3ef)

**Reply by gregsullivan** [reply]
Could you please try deleting the `node_modules` folder and running `npm install` again? That folder should have significantly more subfolders, so something seems to have gone wrong in the installation process.

**Reply by BadScooter1980** [reply]
December 2024 - I can confirm the same issue, have removed the node_modules folder and run npm install several times and get the same, few subfolders. The following npm run dev and npm run watch fail, with the error - sh: run-p: command not found

**Reply by BadScooter1980** [reply]
OK, after some fighting with terminal, I have this working.

I do have one question, though. Is it possible to modify the style.css author section? I have various custom links that I generally add, and have tried to do this but, of course, it is updated each time the css rebuilds.

**Reply by gregsullivan** [reply]
Hi @BadScooter1980—Very glad you sorted this out!

For the benefit of anyone seeing this in the future, do you remember what it was that ultimately resolved the problem? I've never been able to duplicate this or isolate what it is causing people to occasionally encounter this issue.

**Reply by BadScooter1980** [reply]
In all honesty, I can't remember but I know I had to uninstall and reinstall npm a few times and it just sort of worked.

**Reply by gregsullivan** [reply]
Thanks for this—much appreciated!

**Comment by beachdbishop** [comment]
You might try modifying `tailwind/custom/file-header.css` before running a build step.

**Reply by BadScooter1980** [reply]
Super, thank you!

**Comment by coreyp8000** [comment]
For anyone in the future with a similar issue.

I've had this issue before and resolved it by deleting node_modules and redoing `npm install`. This time it didn't work, `npm install` would just add the same few empty subfolders in node_modules.

I resolved it by opening package.json and renaming the "devDependencies" object to "dependencies", deleting the node_modules and redone `npm install`.

All packages are now installed, and the run commands work as normal.

**Reply by gregsullivan** [reply]
Can you please confirm whether you're also using Local, as with some of the other comments above?

If yes, are you using the terminal inside Local, or are you using something like the macOS Terminal app / an application like VS Code?

Regardless of the above, are you syncing up to / down from a host in any way? E.g., via Local, or using something like WP Migrate?

If you or anyone else encounters this issue again, I would be curious to know whether `npm cache clean --force` resolves the issue.

Thanks for posting this!

**Reply by coreyp8000** [reply]
Yes, I am using Local. I'm using Linux and VS Code as my editor. The commands would not work in either my regular terminal app or in VS Code's integrated terminal.

I'm not syncing to any live hosts. I did try `npm cache clean --force` and updating to the latest npm version too, to no success before renaming "devDependencies".

Figuring maybe it is an issue with Linux/MacOS, but I have been able to run the commands first try with no problems on previous projects using both Linux and Windows.

**Reply by gregsullivan** [reply]
I found this on the Local WP forum:

> I’ve started running into an issue that when I use the “Open site shell” feature I can no longer run ‘npm install’ correctly without manually setting the NODE_ENV to ‘development’. (Otherwise it skips devDependencies)

Can you confirm what Local WP is using for `NODE_ENV`?

**Reply by coreyp8000** [reply]
I have no idea what Local WP is using for `NODE_ENV`. I can't find any further information online, .env files, etc.

**Reply by gregsullivan** [reply]
Could you please try running this:

```
node -e "console.log(process.env.NODE_ENV)"
```

And let me know what you get back?

**Reply by gregsullivan** [reply]
I'm considering adding a `preinstall` script that checks the value of `NODE_ENV` and outputs a message if it is set to production. But please let me know about the above, as that would confirm the issue!

**Reply by gregsullivan** [reply]
I've duplicated the issue locally by running `export NODE_ENV=production` and then `npm install`. This leads to the empty directories as described in this thread and elsewhere.

My solution moving forward is going to be to include a `.npmrc` file with `omit = null` as its contents. This prevents `NODE_ENV=production` from omitting any dependencies.

I think the fundamental problem is likely that some combination of Local WP settings causes `NODE_ENV` to be set to `production`, but I think it's unlikely we'll be able to get that fixed. Given that, I think this is the best solution, as I can't think of a situation where someone would want to run `npm install` in a repository generated by _tw and not have all of the dependencies installed.

Thanks for your help with this!

**Reply by coreyp8000** [reply]
Nice, you've got some sort of solution now.

Just so you know, when I do `node -e "console.log(process.env.NODE_ENV)"` it returns 'undefined'.

Maybe something to take into account.

Thanks

**Reply by gregsullivan** [reply]
Thanks for that—much appreciated!

The big change on my side is that it's definitively the `omit` setting causing the problem, as I can toggle that and recreate the issue everyone has been having. `omit` can be set in a number of ways, including via `NODE_ENV`, but setting `omit` to `null` in `.npmrc` should override all of them.

We'll see! If I keep hearing from people after deploying this change, I'll need to reinvestigate. Fingers crossed for now, and thanks for your help!
